### PR TITLE
fix docker restart for hyperv isolation mode (windows)

### PR DESCRIPTION
**Pull request by andrey-ko**
fix for issue https://github.com/moby/moby/issues/35430

In windows hyperv mode, create container fails if container fs is already mounted

### CODE REVIEW

**Comment by thaJeztah** [comment]
ping @johnstep @jhowardmsft PTAL

File: daemon/restart.go [hunk 1]
```diff
@@ -34,11 +35,16 @@ func (daemon *Daemon) ContainerRestart(name string, seconds *int) error {
...
 	// the container when the container stops and then starts
-	// again
-	if err := daemon.Mount(container); err == nil {
-		defer daemon.Unmount(container)
+	// again. Doesn't work on Windows.
+	if runtime.GOOS != "windows" {
+		if err := daemon.Mount(container); err == nil {
+			defer daemon.Unmount(container)
+		} else {
+			logrus.Warningf("can't mount container (%s) fs for faster restart: %v", container.ID, err)
```
Comment (thaJeztah) [inline daemon/restart.go]: I don't think the warning is needed

File: daemon/restart.go [hunk 2]
```diff
@@ -34,11 +35,16 @@ func (daemon *Daemon) ContainerRestart(name string, seconds *int) error {
...
 func (daemon *Daemon) containerRestart(container *container.Container, seconds int) error {
-	// Avoid unnecessarily unmounting and then directly mounting
+
+	// Optimization: avoid unnecessarily unmounting and then directly mounting
 	// the container when the container stops and then starts
-	// again
-	if err := daemon.Mount(container); err == nil {
-		defer daemon.Unmount(container)
+	// again. Doesn't work on Windows.
+	if runtime.GOOS != "windows" {
```
Comment (thaJeztah) [inline daemon/restart.go]: Discussing this in the maintainers meeting; perhaps we should move this logic to https://github.com/moby/moby/blob/master/daemon/daemon.go#L1203-L1227 (which already has some windows exceptions)

**Review by thaJeztah — CHANGES_REQUESTED** [review]

**Comment by jhowardmsft** [comment]
I'm not sure this is an issue??? On RS5, running moby master as of now:

```
PS E:\> docker run -d --isolation=hyperv microsoft/windowsservercore powershell -command sleep 600
605036c8b89ec4188a970ae5f5e4826f5fc0a06a9441cf18327f80ab7198132e
PS E:\> docker ps -a
CONTAINER ID        IMAGE                         COMMAND                  CREATED             STATUS              PORTS               NAMES
605036c8b89e        microsoft/windowsservercore   "powershell -command…"   12 seconds ago      Up 6 seconds                            gracious_herschel
PS E:\> docker restart 60503
60503
PS E:\> docker ps -a
CONTAINER ID        IMAGE                         COMMAND                  CREATED             STATUS              PORTS               NAMES
605036c8b89e        microsoft/windowsservercore   "powershell -command…"   30 seconds ago      Up 3 seconds                            gracious_herschel
PS E:\> sleep 10; docker restart 60503
60503
PS E:\> docker ps -a
CONTAINER ID        IMAGE                         COMMAND                  CREATED             STATUS              PORTS               NAMES
605036c8b89e        microsoft/windowsservercore   "powershell -command…"   59 seconds ago      Up 3 seconds                            gracious_herschel
PS E:\> docker ps -a
CONTAINER ID        IMAGE                         COMMAND                  CREATED              STATUS              PORTS               NAMES
605036c8b89e        microsoft/windowsservercore   "powershell -command…"   About a minute ago   Up 6 seconds                            gracious_herschel
PS E:\> docker ps -a
CONTAINER ID        IMAGE                         COMMAND                  CREATED              STATUS              PORTS               NAMES
605036c8b89e        microsoft/windowsservercore   "powershell -command…"   About a minute ago   Up 7 seconds                            gracious_herschel
PS E:\> docker restart 60503
60503
```

**Comment by jhowardmsft** [comment]
And similarly, it doesn't fail on RS1. So I think there's something else going on if you believe this fix is needed.

**Comment by jhowardmsft** [comment]
Oh right, I see when this fails. Slightly different repro to mine to cause the failure.

```
docker create --name foo --isolation=hyperv microsoft/windowsservercore powershell -command sleep 600
docker restart foo
```

This fix is almost correct. And is the right place to put the fix rather than in `Mount()` itself as it is a restart-specific problem. 

Here's what I would do, including an appropriate comment.

First, add `containertypes "github.com/docker/docker/api/types/container"` to imports in restart.go

```
// containerRestart attempts to gracefully stop and then start the
// container. When stopping, wait for the given duration in seconds to
// gracefully stop, before forcefully terminating the container. If
// given a negative duration, wait forever for a graceful stop.
func (daemon *Daemon) containerRestart(container *container.Container, seconds int) error {

	// Determine isolation. If not specified in the hostconfig, use daemon default.
	actualIsolation := container.HostConfig.Isolation
	if containertypes.Isolation.IsDefault(containertypes.Isolation(actualIsolation)) {
		actualIsolation = daemon.defaultIsolation
	}

	// Avoid unnecessarily unmounting and then directly mounting
	// the container when the container stops and then starts
	// again. We do not do this for Hyper-V isolated containers
	// (implying also on Windows) as the HCS must have exclusive
	// access to mount the containers filesystem inside the utility
	// VM.
	if !containertypes.Isolation.IsHyperV(actualIsolation) {
		if err := daemon.Mount(container); err == nil {
			defer daemon.Unmount(container)
		}
	}
```

**Review by lowenna — CHANGES_REQUESTED** [review]
See comment above about how the fix should look

**Comment by jhowardmsft** [comment]
Closing this one. Carrying in https://github.com/moby/moby/pull/38782 with updated fix.

**Pull request closed without merging**